File: numbertheory/archived/sieve-of-eratosthenes.cpp
#include <bits/stdc++.h>

using namespace std;
#define ll long long

// O(nlog(log(n))).
vector<int> sieve(int n) {
    vector<bool> is_prime(n+1, true);
    vector<int> primes;
    for (int i = 2; i <= n; i++) {
        if (is_prime[i]) {
            primes.push_back(i);
            for (int j = i * i; j <= n; j += i)
                is_prime[j] = false;
        }
    }
    return primes;
}

int main() {
    vector<int> d = sieve(128);
    for (int e : d)
        cout << e << " ";
    cout << endl;
}

File: numbertheory/archived/sieve-of-eratosthenes.py
def sieve(n : int) -> list:
	""" Sieve of Eratosthenes """
	is_prime = [True] * (n + 1)
	primes = []
	for i in range(2, n + 1):
		if (is_prime[i]):
			primes.append(i)
			for j in range(i * i, n + 1, i):
				is_prime[j] = False
	return primes

# saw this implementation on ComputerPhile, but Maximum Recursion Depth... :(
# and the version above is about ~89 times faster
# TODO: Write a Haskell version with a "real infinite" list (I'm still a Haskell newbie LOL)
def successor(x):
    x += 1
    yield x
    yield from successor(x)

def sieve_new(s):
    n = next(s)
    yield n
    yield from sieve_new(i for i in s if i % n != 0)

s = sieve_new(successor(1))
llist = []
for _ in range(128):
    llist.append(next(s))
print(llist)

print(sieve(720))